hwtypes/bit_vector_util.py

```diff
@@ -1,14 +1,146 @@
 import functools as ft
 import itertools as it
+import inspect
+import types
+
 from .bit_vector_abc import InconsistentSizeError
+from .bit_vector_abc import BitVectorMeta, AbstractBitVector, AbstractBit
+
+# used as a tag
+class PolyBase: pass
+
+class PolyType(type):
+    def __getitem__(cls, args):
+        # From a typing perspective it would be better to make select an
+        # argument to init instead of making it a type param.  This would
+        # allow types to be cached etc... However, making it an init arg
+        # means type(self)(val) is no longer sufficient to to cast val.
+        # Instead one would need to write type(self)(val, self._select_) or
+        # equivalent and hence would require a major change in the engineering
+        # of bitvector types (they would need to be aware of polymorphism).
+        # Note we can't cache as select is not necessarily hashable.
+
+        T0, T1, select = args
+
+        if not cls._type_check(T0, T1):
+            raise TypeError(f'Cannot construct {cls} from {T0} and {T1}')
+        if not isinstance(select, AbstractBit):
+            raise TypeError('select must be a Bit')
+        if (T0.get_family() is not T1.get_family()
+            or T0.get_family() is not select.get_family()):
+            raise TypeError('Cannot construct PolyTypes across families')
+
+
+        # stupid generator to make sure PolyBase is not replicated
+        # and always comes last
+        bases = *(b for b in cls._get_bases(T0, T1) if b is not PolyBase), PolyBase
+        class_name = f'{cls.__name__}[{T0.__name__}, {T1.__name__}, {select}]'
+        meta, namespace, _ = types.prepare_class(class_name, bases)
+
+        d0 = dict(inspect.getmembers(T0))
+        d1 = dict(inspect.getmembers(T1))
+
+        attrs = d0.keys() & d1.keys()
+        for k in attrs:
+            if k in {'_info_', '__int__', '__repr__', '__str__'}:
+                continue
+
+            m0 = inspect.getattr_static(T0, k)
+            m1 = inspect.getattr_static(T1, k)
+            namespace[k] = build_VCall(select, m0, m1)
```

Comment:
This seems to be the magic, a PolyType store's it's select condition, operators are dispatched based on that condition when invoked.

Reply:
In theory this might work with magma types?

Reply:
Actually not quite, Poly would need to implement the magma protocol, but then I think it could work, maybe just by concatenating the bits of the two subtypes together as the bits representation?

Reply:
I plan on handling magma with my MagmaBV proposal. https://github.com/phanrahan/magma/issues/587

MagmaBV will use the same machinery to build ites as BitVector / SMTBitVector and implement the magma protocol.